ext/dcos-installer/dcos_installer/action_lib.py

```diff
@@ -43,27 +43,73 @@ class ExecuteException(Exception):
+    assert step in ['preflight', 'deploy', 'postflight']
+    step += '.json'
+
+    step_path = os.path.join(state_json_dir, step)
+    step_json = _read_state_file(step_path)
+    if not step_json:
+        # TODO: Fix exception here
+        raise Exception("Could not get hosts from {}".format(step_path))
+
+    assert 'hosts' in step_json, "Field `hosts` not found, location {}".format(step_path)
+
+    nodes = []
+    for node_ip, node_params in step_json['hosts'].items():
+        assert 'host_status' in node_params, "Field `host_status` not found for node {}".format(node_ip)
+        assert 'tags' in node_params, 'Field `tags` not found for node {}'.format(node_ip)
+        if node_params['host_status'] == 'success':
+            nodes.append(Node(node_ip, tags=node_params['tags']))
+        else:
+            assert 'role' in node_params['tags'], 'tags must contain parameter role. Got {}'.format(node_params['tags'])
+            if node_params['tags']['role'] == 'master':
+                raise Exception("Master node {} has failed. Uable to continue".format(node_ip))
+            log.debug("node {} unhealthy {}".format(node_ip, node_params['host_status']))
+
+    assert nodes, "No healthy nodes found"
+    return nodes
+
+
+def nodes_count_by_type(targets):
+    assert isinstance(targets, list)
+    total_masters = 0
+    total_agents = 0
+    for target in targets:
+        assert isinstance(target, Node)
+        assert 'role' in target.tags
+        if target.tags['role'] == 'master':
+            total_masters += 1
+        else:
+            total_agents += 1
+
     return {
-        'total_masters': len(config['master_list']),
-        'total_agents': total_agents_count
+        'total_masters': total_masters,
+        'total_agents': total_agents
     }
 
 
 def get_full_nodes_list(config):
-    def add_nodes(nodes, tag):
-        return [Node(node, tag) for node in nodes]
-
-    node_role_map = {
-        'master_list': 'master',
-        'agent_list': 'agent',
-        'public_agent_list': 'public_agent'
+    role_params = {
+        'master': {
+            'tags': {'role': 'master', 'dcos_install_param': 'master'},
+            'hosts': config['master_list']
+        },
+        'agent': {
+            'tags': {'role': 'agent', 'dcos_install_param': 'slave'},
+            'hosts': config.get('agent_list', [])
+        },
+        'public_agent': {
+            'tags': {'role': 'public_agent', 'dcos_install_param': 'slave_public'},
+            'hosts': config.get('public_agent_list', [])
+        }
     }
+    ssh_port = config.get('ssh_port', 22)
+
     full_target_list = []
-    for config_field, role in node_role_map.items():
-        if config_field in config:
-            full_target_list += add_nodes(config[config_field], {'role': role})
+    for role, params in role_params.items():
+        full_target_list += [Node('{}:{}'.format(node, ssh_port), params['tags']) for node in params['hosts']]
```

Comment:
Seems a bit silly for us to squash these together than parse them apart then inside Node(). Could we make node take the ip and port separately?

This I think also changes what we allow here. Before, if the nodes happened to be of the form `host:port` things would work, now they don't (We always force the global ssh_port)